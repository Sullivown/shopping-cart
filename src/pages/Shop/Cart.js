import React, { useEffect } from 'react';
import updateDocumentTitle from '../../helpers/updateDocumentTitle';
import CartProductPreview from '../../components/CartProductPreview';

function Cart(props) {
	useEffect(() => {
		updateDocumentTitle(props.title);
	});

	const cartProductElements = props.cart.map((item, index) => (
		<CartProductPreview
			key={index}
			product={item.item}
			quantity={item.quantity}
			addToCart={props.addToCart}
		/>
	));

	return (
		<div className='cart'>
			<h2>Cart</h2>
			<div className='cart-details'>
				<div className='cart-products-list'>{cartProductElements}</div>
				<div className='cart-summary'>
					<h3>Cart Total</h3>
					<p>£££</p>
					<button>Checkout</button>
				</div>
			</div>
		</div>
	);
}

export default Cart;
