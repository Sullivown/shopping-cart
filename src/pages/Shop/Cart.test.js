import React from 'react';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import { BrowserRouter as Router } from 'react-router-dom';
import TestRenderer from 'react-test-renderer';

import Cart from './Cart';
import userEvent from '@testing-library/user-event';

const mockCart = [
	{
		item: {
			id: 1,
			title: 'Product 1',
			price: 2.99,
			image: jest.mock('../../assets/images/product-images/1.png'),
		},
		quantity: 1,
	},
	{
		item: {
			id: 2,
			title: 'Product 2',
			price: 10,
			image: jest.mock('../../assets/images/product-images/2.png'),
		},
		quantity: 2,
	},
	{
		item: {
			id: 3,
			title: 'Product 3',
			price: 5.99,
			image: jest.mock('../../assets/images/product-images/3.png'),
		},
		quantity: 3,
	},
];

describe('Cart component tests', () => {
	const addToCartMock = jest.fn();

	it('renders', () => {
		const cartRenderer = TestRenderer.create(
			<Router>
				<Cart title='Cart' cart={mockCart} addToCart={addToCartMock} />
			</Router>
		).toJSON();

		expect(cartRenderer).toMatchSnapshot();
	});

	it('renders empty cart (0 items) correctly', () => {
		render(
			<Router>
				<Cart title='Cart' cart={[]} addToCart={addToCartMock} />
			</Router>
		);

		const cartItems = screen.getByTestId('cart-products-list');

		expect(cartItems.childNodes.length).toBe(1);
		expect(cartItems.childNodes[0].textContent).toBe('Your cart is empty!');
	});

	it('renders 3 cart items', () => {
		render(
			<Router>
				<Cart title='Cart' cart={mockCart} addToCart={addToCartMock} />
			</Router>
		);

		const cartItems = screen.getByTestId('cart-products-list');

		expect(cartItems.childNodes.length).toBe(3);
	});

	it('calls addToCart correctly when qty edited and submitted', () => {
		const testCart = [mockCart[0]];
		render(
			<Router>
				<Cart title='Cart' cart={testCart} addToCart={addToCartMock} />
			</Router>
		);

		const qtyInput = screen.getByRole('spinbutton', { type: 'number' });

		userEvent.clear(qtyInput);
		userEvent.type(qtyInput, '0{Enter}');

		expect(qtyInput.value).toBe('0');
		expect(addToCartMock).toBeCalledWith(testCart[0].item, 0, true);
	});
});
