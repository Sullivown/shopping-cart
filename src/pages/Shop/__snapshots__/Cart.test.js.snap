// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart component tests renders 1`] = `
<div
  className="page-content cart"
>
  <h2>
    Cart
  </h2>
  <div
    className="cart-details"
  >
    <div
      className="cart-products-list"
      data-testid="cart-products-list"
    >
      <section
        className="cart-product-preview"
      >
        <img
          alt="Product 1"
          onClick={[Function]}
          src={
            Object {
              "advanceTimersByTime": [Function],
              "advanceTimersToNextTimer": [Function],
              "autoMockOff": [Function],
              "autoMockOn": [Function],
              "clearAllMocks": [Function],
              "clearAllTimers": [Function],
              "createMockFromModule": [Function],
              "deepUnmock": [Function],
              "disableAutomock": [Function],
              "doMock": [Function],
              "dontMock": [Function],
              "enableAutomock": [Function],
              "fn": [Function],
              "genMockFromModule": [Function],
              "getRealSystemTime": [Function],
              "getTimerCount": [Function],
              "isMockFunction": [Function],
              "isolateModules": [Function],
              "mock": [Function],
              "mocked": [Function],
              "requireActual": [Function],
              "requireMock": [Function],
              "resetAllMocks": [Function],
              "resetModules": [Function],
              "restoreAllMocks": [Function],
              "retryTimes": [Function],
              "runAllImmediates": [Function],
              "runAllTicks": [Function],
              "runAllTimers": [Function],
              "runOnlyPendingTimers": [Function],
              "setMock": [Function],
              "setSystemTime": [Function],
              "setTimeout": [Function],
              "spyOn": [Function],
              "unmock": [Function],
              "unstable_mockModule": [Function],
              "useFakeTimers": [Function],
              "useRealTimers": [Function],
            }
          }
        />
        <h3
          onClick={[Function]}
        >
          Product 1
        </h3>
        <input
          min="0"
          onBlur={[Function]}
          onChange={[Function]}
          onKeyUp={[Function]}
          type="number"
          value={1}
        />
        <div>
          £2.99
        </div>
      </section>
      <section
        className="cart-product-preview"
      >
        <img
          alt="Product 2"
          onClick={[Function]}
          src={
            Object {
              "advanceTimersByTime": [Function],
              "advanceTimersToNextTimer": [Function],
              "autoMockOff": [Function],
              "autoMockOn": [Function],
              "clearAllMocks": [Function],
              "clearAllTimers": [Function],
              "createMockFromModule": [Function],
              "deepUnmock": [Function],
              "disableAutomock": [Function],
              "doMock": [Function],
              "dontMock": [Function],
              "enableAutomock": [Function],
              "fn": [Function],
              "genMockFromModule": [Function],
              "getRealSystemTime": [Function],
              "getTimerCount": [Function],
              "isMockFunction": [Function],
              "isolateModules": [Function],
              "mock": [Function],
              "mocked": [Function],
              "requireActual": [Function],
              "requireMock": [Function],
              "resetAllMocks": [Function],
              "resetModules": [Function],
              "restoreAllMocks": [Function],
              "retryTimes": [Function],
              "runAllImmediates": [Function],
              "runAllTicks": [Function],
              "runAllTimers": [Function],
              "runOnlyPendingTimers": [Function],
              "setMock": [Function],
              "setSystemTime": [Function],
              "setTimeout": [Function],
              "spyOn": [Function],
              "unmock": [Function],
              "unstable_mockModule": [Function],
              "useFakeTimers": [Function],
              "useRealTimers": [Function],
            }
          }
        />
        <h3
          onClick={[Function]}
        >
          Product 2
        </h3>
        <input
          min="0"
          onBlur={[Function]}
          onChange={[Function]}
          onKeyUp={[Function]}
          type="number"
          value={2}
        />
        <div>
          £20.00
        </div>
      </section>
      <section
        className="cart-product-preview"
      >
        <img
          alt="Product 3"
          onClick={[Function]}
          src={
            Object {
              "advanceTimersByTime": [Function],
              "advanceTimersToNextTimer": [Function],
              "autoMockOff": [Function],
              "autoMockOn": [Function],
              "clearAllMocks": [Function],
              "clearAllTimers": [Function],
              "createMockFromModule": [Function],
              "deepUnmock": [Function],
              "disableAutomock": [Function],
              "doMock": [Function],
              "dontMock": [Function],
              "enableAutomock": [Function],
              "fn": [Function],
              "genMockFromModule": [Function],
              "getRealSystemTime": [Function],
              "getTimerCount": [Function],
              "isMockFunction": [Function],
              "isolateModules": [Function],
              "mock": [Function],
              "mocked": [Function],
              "requireActual": [Function],
              "requireMock": [Function],
              "resetAllMocks": [Function],
              "resetModules": [Function],
              "restoreAllMocks": [Function],
              "retryTimes": [Function],
              "runAllImmediates": [Function],
              "runAllTicks": [Function],
              "runAllTimers": [Function],
              "runOnlyPendingTimers": [Function],
              "setMock": [Function],
              "setSystemTime": [Function],
              "setTimeout": [Function],
              "spyOn": [Function],
              "unmock": [Function],
              "unstable_mockModule": [Function],
              "useFakeTimers": [Function],
              "useRealTimers": [Function],
            }
          }
        />
        <h3
          onClick={[Function]}
        >
          Product 3
        </h3>
        <input
          min="0"
          onBlur={[Function]}
          onChange={[Function]}
          onKeyUp={[Function]}
          type="number"
          value={3}
        />
        <div>
          £17.97
        </div>
      </section>
    </div>
    <div
      className="cart-summary"
    >
      <h3>
        Cart Total
      </h3>
      <p>
        £40.96
      </p>
      <button>
        Checkout
      </button>
    </div>
  </div>
</div>
`;
